libs/ledgerjs/packages/cryptoassets/src/crypto-assets-importer/fetch/index.ts

```diff
@@ -1,13 +1,87 @@
+import { getEnv } from "@ledgerhq/live-env";
+import { CryptoCurrencyId } from "@ledgerhq/types-cryptoassets";
 import axios, { AxiosError } from "axios";
 
-const CRYPTO_ASSETS_BASE_URL = "https://cdn.live.ledger.com/cryptoassets";
-
-export const fetchTokens = async <T>(filename: string): Promise<[T, string | undefined]> => {
+export const fetchTokensFromCDN = async <T>(filename: string): Promise<[T, string | undefined]> => {
   try {
-    const { data, headers } = await axios.get<T>(`${CRYPTO_ASSETS_BASE_URL}/${filename}`);
+    const { data, headers } = await axios.get<T>(`${getEnv("DYNAMIC_CAL_BASE_URL")}/${filename}`);
     return [data, headers.etag];
   } catch (err) {
     const error = err as AxiosError;
-    throw new Error(`${error.message} ${CRYPTO_ASSETS_BASE_URL}/${filename}`);
+    throw new Error(`${error.message} ${getEnv("DYNAMIC_CAL_BASE_URL")}/${filename}`);
+  }
+};
+
+export type CALServiceOutput = {
+  type: string;
+  id: string;
+  blockchain_name: string;
+  chain_id: number;
+  contract_address: string;
+  token_identifier: string;
+  decimals: number;
+  delisted: boolean;
+  network_type: string;
+  name: string;
+  symbol: string;
+  ticker: string;
+  units: string;
+  live_signature: string;
+  exchange_app_config_serialized: string;
+  exchange_app_signature: string;
+};
+
+export const fetchTokensFromCALService = async <T extends Array<keyof CALServiceOutput>>(
+  chainDetails: {
+    chainId?: string | number;
+    standard?: string;
+    blockchain_name?: CryptoCurrencyId;
+  },
+  output: T,
+  etag?: string | null,
+  next?: { cursor: string; tokens: Pick<CALServiceOutput, T[number]>[] },
+): Promise<{ tokens: Pick<CALServiceOutput, T[number]>[]; hash: string | undefined }> => {
+  try {
+    const { data, headers } = await axios.get<Pick<CALServiceOutput, T[number]>[]>(
+      `${getEnv("CAL_SERVICE_URL")}/v1/tokens`,
+      {
+        params: {
+          ...(next?.cursor ? { cursor: next.cursor } : {}),
+          chain_id: chainDetails.chainId,
+          standard: chainDetails.standard,
+          blockchain_name: chainDetails.blockchain_name,
+          output: output.join(),
+        },
+        headers: etag
+          ? {
+              "If-None-Match": etag,
+            }
+          : undefined,
+      },
+    );
+
+    const cursor = headers["x-ledger-next"];
```

Comment:
just a question, what is this "cursor"? 
is it for pagination? If so, should we make a loop to fetch all pages of tokens?

Reply:
Yes a cursor in like a pointer to a specific db record, and that's used for pagination indeed.
And there is already a loop to fetch all the tokens as if you do have a cursor, the function will call itself again with the cursor this time to continue the request